## Объектно-ориентированное программирование и информационная безопасность

*Валерий Семенов, Самарский университет*

<div style="text-align:center"><img src="Python.png" width="200"></div>

<h1 style="text-align:center; margin-top:40px">Вычисления в Python</h1>

<div style="text-indent:30px; text-align:justify">Привычные арифметические действия (сложение, вычитание, умножение, деление) в Python выглядят так же, как и в обычных калькуляторах:</div>

In [1]:
1 + 4  # сложение

5

In [2]:
5 * 4 - 9  # умножение и вычитание

11

In [3]:
7 / 2  # деление

3.5

<div style="text-indent:30px; text-align:justify">Однако с делением все не так просто. Python всегда будет выдавать результат в виде числа с плавающей десятичной точкой (<strong>float</strong>), даже тогда, когда ожидается целочисленный ответ.</div>
<div style="text-indent:30px; text-align:justify; margin-top:20px">Например:</div>

In [4]:
8 / 2  # не 4

4.0

<div style="text-indent:30px; text-align:justify">Получился дробный результат, где дробная часть равна 0.</div>
<div style="text-indent:30px; text-align:justify">Как быть, если нужен ответ в виде целого числа?</div>

<div style="text-indent:30px; text-align:justify; margin-top:20px">Можно воспользоваться целочисленным делением.</div>

In [5]:
8 // 2  # теперь 4

4

<div style="text-indent:30px; text-align:justify">Тут важно помнить, что при использовании оператора <strong>//</strong> дробная часть всегда будет просто отбрасываться – никакого округления происходить не будет. Для округления (обычное арифметическое, в большую и в меньшую сторону) существуют специальные функции, и мы их обсудим позже.</div>

In [6]:
7 // 2  # от 3.5 осталось 3

3

<div style="text-indent:30px; text-align:justify">Что еще можно делать с числами?</div>
<div style="text-indent:30px; text-align:justify">Возводить в степень и извлекать из них корень.</div>
    
<div style="text-indent:30px; text-align:justify">При расчетах на калькуляторе (и в Excel) для возведения числа в степень мы обычно используем символ <strong>^</strong>.</div> 

<div style="text-indent:30px; text-align:justify; margin-top:20px">Попробуем:</div>

In [7]:
6 ^ 2

4

<div style="text-indent:30px; text-align:justify">Получилось что-то неожиданное!</div>
<div style="text-indent:30px; text-align:justify; margin-top:20px">Дело в том, что в Python оператор <strong>^</strong> используется для побитного сложения по модулю два.</div>
<div style="text-indent:30px; text-align:justify; margin-top:20px">Для возведения числа в степень потребуется <strong>**</strong> (<i>смотрите картинку "Математические операторы" из предыдущего блокнота</i>):</div>

In [8]:
6 ** 2  # как нужно

36

In [9]:
6 ** 3

216

In [10]:
6 ** (1/3)  # дробная степень - корень 3 степени

1.8171205928321397

<div style="text-indent:30px; text-align:justify; margin-top:20px">В Python есть небольшое количество встроенных базовых функций, в том числе математических:</div>

In [11]:
a = abs(-5.3459)  # возвращает абсолютное значение числа
b = round(a, 2)  # округляет число до n знаков после запятой (n — необязательный аргумент)
a, b

(5.3459, 5.35)

<div style="text-indent:30px; text-align:justify">Теперь попробуем извлечь квадратный корень из числа с помощью привычной функции <strong>sqrt()</strong> (C++, Java, JavaScript, Turbo Pascal):</div>

In [12]:
sqrt(9)  # не получается!

NameError: name 'sqrt' is not defined

<div style="text-indent:30px; text-align:justify">Python пишет, что не знает, что такое <strong>sqrt</strong>. </div>
    
<div style="text-indent:30px; text-align:justify; margin-top:20px">В каких случаях Python может такое писать? Например, если мы опечатались в названии функции (Python не понимает, что мы от него хотим) или если мы пытаемся обратиться к функции, которая не является базовой (Python не знает, откуда ее брать). В данном случае мы столкнулись со второй проблемой. </div>

<h2 style="text-align:center; margin-top:40px">Модуль <strong>math</strong></h2>
    
<div style="text-indent:30px; text-align:justify">Функция для вычисления квадратного корня из числа хранится в специальном модуле <strong>math</strong>. Этот модуль стандартный, дополнительно устанавливать его не нужно. Но для того чтобы воспользоваться этой функцией, нужно сначала импортировать модуль, а потом вызвать из него функцию <strong>sqrt()</strong>.</div>

In [13]:
import math   # импортируем модуль maths
math.sqrt(9)  # теперь все работает

3.0

<div style="text-indent:30px; text-align:justify">Если из <strong>math</strong> нам нужна только одна функция <strong>sqrt()</strong>, можно извлечь только ее, и тогда прописывать название модуля перед функцией не понадобится:</div>

In [14]:
from math import sqrt
sqrt(16)    # так тоже работает

4.0

<div style="text-indent:30px; text-align:justify">Иногда, если неизвестно, сколько функций потребуется, но каждый раз прописывать полностью название модуля не хочется, можно импортировать его сразу с сокращенным названием (псевдонимом):</div>

In [15]:
import math as ma   # сократили название модуля math до ma
ma.sqrt(9)

3.0

<div style="text-indent:30px; text-align:justify">Примечание: то же будет верно и для библиотек. Пока мы столкнулись с двумя модулями – <strong>os</strong> и <strong>math</strong>, в темах по обработке данных мы будем активно работать с разными библиотеками. Если совсем упростить, модуль – это набор функций, а библиотека – это набор модулей, то есть что-то более сложное по структуре и более насыщенное по функционалу.</div>

<div style="text-indent:30px; text-align:justify; margin-top:20px">В модуле <strong>math</strong> есть много полезных функций для вычислений. Чтобы посмотреть, какие функции там есть, после импортирования модуля можно набрать <strong>dir(math)</strong>:</div>

In [16]:
import math
dir(math)

['__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'cbrt',
 'ceil',
 'comb',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'dist',
 'e',
 'erf',
 'erfc',
 'exp',
 'exp2',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'isqrt',
 'lcm',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'nextafter',
 'perm',
 'pi',
 'pow',
 'prod',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'sumprod',
 'tan',
 'tanh',
 'tau',
 'trunc',
 'ulp']

<div style="text-indent:30px; text-align:justify; margin-top:20px">Примеры:</div>

In [17]:
math.log(2)   # натуральный логарифм

0.6931471805599453

In [18]:
math.log10(100)   # десятичный логарифм (логарифм по основанию 10)

2.0

In [19]:
math.sin(2*3.14)   # синус

-0.0031853017931379904

<div style="text-indent:30px; text-align:justify; margin-top:20px">Здесь же хранятся функции для округления в большую или меньшую сторону:</div>

In [20]:
math.ceil(8.7)   # ceil - потолок

9

In [21]:
math.floor(8.7)  # floor - пол

8

<div style="text-indent:30px; text-align:justify; margin-top:20px">А еще из модуля <strong>math</strong> можно импортировать константы <strong>π</strong>, <strong>e</strong> и экспоненциальную функцию <strong>exp()</strong>:</div>

In [22]:
from math import pi, exp, e   # можно сразу несколько - перечислить через запятую

In [23]:
pi

3.141592653589793

In [24]:
e

2.718281828459045

In [25]:
exp

<function math.exp(x, /)>

In [26]:
exp(1)   # e = e^1 = exp(1)

2.718281828459045

<div style="text-indent:30px; text-align:justify; margin-top:20px">Примечание: если мы знаем, что будем использовать практически все функции из модуля <strong>math</strong>, их можно извлечь все сразу, и тогда прописывать название библиотеки нам не понадобится нигде:</div>

In [27]:
from math import *

<div style="text-indent:30px; text-align:justify">Однако в общем случае так делать нежелательно, потому что это нерационально и увеличивает время исполнения кода (Python подгружает все функции, а мы пользуемся только двумя, например).</div>
<div style="text-indent:30px; text-align:justify">Также мы можем столкнуться с еще одной неприятностью. Например, мы используем свою переменную, имя которой совпадает с именем константы или функции из модуля:</div>

In [28]:
gamma = 12345
gamma

12345

In [29]:
from math import *
gamma

<function math.gamma(x, /)>

<div style="text-indent:30px; text-align:justify">Рассмотрим еще один интересный момент, связанный с вычислениями в Python:</div>

In [30]:
1 / 18 ** 25

4.1513310942010236e-32

<div style="text-indent:30px; text-align:justify">Результат выше это <a href="https://ru.wikipedia.org/wiki/%D0%A7%D0%B8%D1%81%D0%BB%D0%BE_%D1%81_%D0%BF%D0%BB%D0%B0%D0%B2%D0%B0%D1%8E%D1%89%D0%B5%D0%B9_%D0%B7%D0%B0%D0%BF%D1%8F%D1%82%D0%BE%D0%B9"><strong>число с плавающей точкой</strong></a>, то есть компьютерная форма экспоненциальной записи числа. Возможно, тот, кто считал что-то на научных или инженерных калькуляторах, уже сталкивался с такой записью. Здесь <strong>e-32</strong> – это <strong>10<sup>−32</sup></strong>, а вся запись означает <strong>4.1513310942010236&bull;10<sup>−32</sup></strong>, то есть примерно <strong>4.15&bull;10<sup>−32</sup></strong>.</div>

<div style="text-indent:30px; text-align:justify">Теоретически, если число было очень большим, после <strong>e</strong> стояла бы положительная степень. Но в Python такое не случается, обычно он выводит огромные числа, просто переходя на новую строку, если места на одной не хватает.</div>

In [31]:
33 ** 490

1179860169468393696094850428398574326812951635768632922638247087305169015296897617090552227029728639332790976165833425012873149029569699243476515762146016403288859933975483731219169926506009362673657387238695354440796670371519895159054179885491366910550124753230630656795784813304948275683081727373233853822010494533842659031456932120263703593279055881044608336618479280113137341318187829384933999323035259811785124498715761011415678013143401952409490509789109658195124702004920241991962260580648931584784740866102626676425599544913867187349914307227741835534044125794437572270180201766008780256272679995970685845474461023887843627618537419795424357371662835493400389096889848144392669284697565958297091399784816195776935986604336499743558218049

<div style="text-indent:30px; text-align:justify">Компьютерная форма записи числа отчасти помогает понять, почему дробные числа называются числами с плавающей точкой (<strong>float</strong>). </div>
    
<div style="text-indent:30px; text-align:justify; margin-top:20px">Возьмем число попроще, например, <strong>12.34</strong>.
<div style="text-indent:30px; text-align:justify">Его можно записать как <strong>12.34</strong>, как <strong>1.234&bull;10</strong>, как <strong>123.4&bull;10<sup>−1</sup></strong>, <strong>1234&bull;10<sup>−2</sup></strong> и так далее. Точка, отделяющая дробную часть от целой, будет «плавать», однако само число при этом меняться не будет, будут меняться только множители – разные степени десятки.</div>

<h3 style="text-align:center; margin-top:40px">Переменные в Python — это указатели (ссылки)</h3>
    
<div style="text-indent:30px; text-align:justify">Переменная в Python является лишь ссылкой на объект в памяти. При создании любой переменной (число, строка или массив) в нее записывается ссылка на объект, а сам объект находится где-то в оперативной памяти далеко от самой переменной со ссылкой. Таким образом, несколько переменных могут указывать на один объект, и при изменении объекта (например, списка) изменится результат обращения к нему с использованием каждой переменной.</div>

<div style="text-indent:30px; text-align:justify">Все объекты в Python имеют свой уникальный идентификатор (номер объекта). Идентификатор присваивается объекту при его создании. Идентификатор является адресом памяти объекта и будет отличаться при каждом запуске программы. </div>

<div style="text-indent:30px; text-align:justify; margin-top:20px">Чтобы получить уникальный идентификатор объекта в программе используется функция <strong>id(object)</strong>:</div>

In [32]:
a = 4
b = 5
a, b

(4, 5)

In [33]:
id(a), id(b)  # Посмотрим идентификаторы переменных a и b

(140725525555736, 140725525555768)

In [34]:
a = b
a, b

(5, 5)

In [35]:
id(a), id(b)  # Посмотрим идентификаторы переменных a и b

(140725525555768, 140725525555768)

<div style="text-indent:30px; text-align:justify">Теперь переменные <strong>a</strong> и <strong>b</strong> ссылаются на один и тот же объект:</div>

<div style="text-align:center"><img src="id.png" width="400"></div>

<h1 style="text-align:center; margin-top:60px">Типы данных в Python</h1>

<div style="text-indent:30px; text-align:justify">Типом данных называют большое количество значений и комплекс операций, которые можно использовать на этих значениях. В языке программирования Python к ним относятся числа, строки, списки, словари, кортежи, множества и логический тип данных. Типы данных в Python можно разделить на:</div>
<div style=" margin-left:50px">
<li style="text-align:justify">изменяемые (множества, списки, словари);</li>
<li style="text-align:justify">неизменяемые (кортежи, строки, числа);</li>
<li style="text-align:justify">упорядоченные (списки, строки, кортежи, словари);</li>
<li style="text-align:justify">неупорядоченные (множества).</li>
</div>

<div style="text-align:center"><img src="data_types.png"></div>

<div style="text-indent:30px; text-align:justify"><strong>Последовательности</strong> – упорядоченные коллекции элементов. В последовательностях каждый элемент имеет свою позицию (номер, индекс) в соответствии с которой элемент сохраняется и извлекается из последовательности.</div>  
<div style="text-indent:30px; text-align:justify">Элементы <strong>отображений</strong> являются соответствиями одних данных другим. Первые при этом называют ключами, вторые – значениями. </div>

<h2 style="text-align:center; margin-top:60px">Строки (<strong>str</strong>)</h2>

In [36]:
s = "Python3"
s

'Python3'

In [37]:
print(s)

Python3


In [38]:
type(s)

str

In [39]:
s += s
s

'Python3Python3'

In [40]:
# Тройные кавычки позволяют создавать многострочные строки 
txt = '''Трагедия Пушкина «Моцарт и Сальери» занимает всего десять страниц. О чем она? 
О зависти или о том, что «гений и злодейство — две вещи несовместные»? 
Есть ли оправдание Сальери, который, по версии Пушкина, отравил Моцарта?'''
txt

'Трагедия Пушкина «Моцарт и Сальери» занимает всего десять страниц. О чем она? \nО зависти или о том, что «гений и злодейство — две вещи несовместные»? \nЕсть ли оправдание Сальери, который, по версии Пушкина, отравил Моцарта?'

In [41]:
print(txt)

Трагедия Пушкина «Моцарт и Сальери» занимает всего десять страниц. О чем она? 
О зависти или о том, что «гений и злодейство — две вещи несовместные»? 
Есть ли оправдание Сальери, который, по версии Пушкина, отравил Моцарта?


In [42]:
s = 'a'
s

'a'

In [43]:
SSS = "Самарский университет"
print(SSS)

Самарский университет


In [44]:
SSS = ""Самарский университет""
print(SSS)

SyntaxError: invalid syntax (228149484.py, line 1)

In [45]:
SSS = """Самарский университет"""
print(SSS)

Самарский университет


<h3 style="text-align:center; margin-top:20px">Обращение по индексу</h3>

![img](https://python-tricks.com/wp-content/uploads/2019/08/Positive-and-Negative-Indexing.jpg)

<div style="text-indent:30px; text-align:justify">Чтобы обратиться к элементу по индексу используют запись <strong>имя_последовательности[индекс]</strong> </div> 
<div style="text-indent:30px; text-align:justify">Индексация в Python начинается с <strong>0</strong>. Также можно использовать обратную индексацию - номер последнего элемента имеет индекс <strong>-1</strong>.</div>

In [46]:
s = 'Python'

In [47]:
s[0]

'P'

In [48]:
s[-1]

'n'

In [49]:
s[5]

'n'

In [50]:
s[3]

'h'

<h3 style="text-align:center; margin-top:20px">Срезы</h3>

<div style="text-indent:30px; text-align:justify">Операция взятия среза: <strong>имя_последовательности[</strong>начало<strong>:</strong>конец <i>(не включается)</i><strong>:</strong>шаг<strong>]</strong>.</div>  
<div style="text-indent:30px; text-align:justify">Если начало не указано - используется 0; если не указан конец, то берется конец строки; если не указан шаг, используется шаг равный 1.</div>

In [51]:
s = 'Python'

In [52]:
s[:4]

'Pyth'

In [53]:
s[4:]

'on'

In [54]:
s[1:4]

'yth'

In [55]:
s[::2]

'Pto'

In [56]:
s[:]

'Python'

In [57]:
'str' + 'ing' # конкатенация (склеивание)

'string'

In [58]:
'Py'*3 # повторение

'PyPyPy'

In [59]:
"Py" in "Python"

True

In [60]:
"2" not in 'Python'

True

<div style="text-indent:30px; text-align:justify">Строки - неизменяемый тип данных!</div>

In [1]:
a = 'Привет'
a[-1] = 'д' # Строки - неизменяемый тип данных!

TypeError: 'str' object does not support item assignment

In [62]:
a = a[0:-1] + 'д'
a

'Привед'

In [63]:
a = 'Приветики'
a

'Приветики'

In [64]:
txt = "Универ"
txt = txt[:2] + "ИВ" + txt[4:]
txt

'УнИВер'

<div style="text-indent:30px; text-align:justify">Для преобразования объектов других типов в строки используется функция <strong>str()</strong>:</div>

In [65]:
s1 = 8765.56
s2 = str(s1)
print(s1, s2)
print(type(s1), type(s2))

8765.56 8765.56
<class 'float'> <class 'str'>
